csharp/src/Google.Protobuf/JsonParser.cs

```diff
@@ -67,15 +67,18 @@ public sealed class JsonParser
 
         private static readonly JsonParser defaultInstance = new JsonParser(Settings.Default);
 
-        private static readonly Dictionary<string, Action<JsonParser, IMessage, JsonTokenizer>>
-            WellKnownTypeHandlers = new Dictionary<string, Action<JsonParser, IMessage, JsonTokenizer>>
+        // TODO: Consider introducing a class containing parse state of the parser, tokenizer and depth. That would simplify these handlers
```

Comment:
I wondered about that, in terms of getting it absolutely for free by the count of the stack. But yes, we could keep a separate counter in the tokenizer. I *think* Struct is already okay, but wrapper types wouldn't count as much as they logically should in the current implementation.

Shall I make that change as a second commit in this PR, so you can see what it's like and we can roll it back easily if we decide not to do it?